File: leetcode/medium/39-combo-sum/dfs.py
"""
  @author Vortexx2
  Problem 39 - Combination Sum

  Runtime - 48 ms
  Memory Usage - 14.5 MB
"""
from typing import List


class Solution:
  def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
    candidates = sorted(candidates)
    res = []

    def dfs(target: int, path: List[int], index: int):
      if target == 0:
        res.append(path)
        return

      for i in range(index, len(candidates)):
        if candidates[i] > target:
          return

        dfs(target - candidates[i], path + [candidates[i]], i)

    dfs(target, [], 0)
    return res

File: leetcode/medium/39-combo-sum/dfs.cpp
/**
 * @file dfs.cpp
 * @author Vortexx2
 * @brief Problem 39 - Combination Sum
 * @date 12-08-2021
 * 
 * Runtime - 0 ms
 * Memory Usage - 11.2 MB
 */
#include <iostream>
#include <vector>
#include <deque>
#include <algorithm>
#include <list>
#include <unordered_map>
#include <map>
#include <climits>
#include <set>
#include <stack>
#include <queue>
#include <string>

using namespace std;

class Solution
{
public:
  vector<vector<int>> combinationSum(vector<int> &candidates, int target)
  {
    vector<int> path;
    vector<vector<int>> res;
    sort(candidates.begin(), candidates.end());

    dfs(candidates, target, path, 0, res);
    return res;
  }

  void dfs(vector<int> &cand, int target, vector<int> &path, int index, vector<vector<int>> &res)
  {
    if (!target)
    {
      res.push_back(path);
      return;
    }

    vector<int> co;
    for (int i = index; i < cand.size(); i++)
    {
      if (target < cand[i])
        return;

      co = path;
      co.push_back(cand[i]);

      dfs(cand, target - cand[i], co, i, res);
    }
  }
};